src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs

```diff
@@ -263,28 +269,59 @@ public ProjectInfo CreateProjectInfoForCurrentState()
         {
             ValidateReferences();
 
+            ImmutableArray<DocumentInfo> documents;
+            ImmutableArray<DocumentInfo> additionalDocuments;
+            ImmutableArray<ProjectReference> projectReferences;
+            ImmutableArray<MetadataReference> metadataReferences;
+            ImmutableArray<AnalyzerReference> analyzerReferences;
+            CompilationOptions compilationOptions;
+            ParseOptions parseOptions;
+            ProjectId id;
+            VersionStamp version;
+            string displayName, assemblyName, filePath, outputFilePath, language;
+            bool hasAllInformation;
+
+            lock (_gate)
+            {
+                documents = _documents.Values.Select(d => d.GetInitialState()).ToImmutableArray();
+                additionalDocuments = _additionalDocuments.Values.Select(d => d.GetInitialState()).ToImmutableArray();
+                metadataReferences = _metadataReferences.Select(r => r.CurrentSnapshot).ToImmutableArray<MetadataReference>();
+                projectReferences = _projectReferences.ToImmutableArray();
+                analyzerReferences = _analyzers.Values.Select(a => a.GetReference()).ToImmutableArray();
+                compilationOptions = this.CurrentCompilationOptions;
+                parseOptions = this.CurrentParseOptions;
+                id = this.Id;
+                version = this.Version;
+                displayName = this.DisplayName;
+                assemblyName = this.AssemblyName ?? this.ProjectSystemName;
+                filePath = this.ProjectFilePath;
+                outputFilePath = this.ObjOutputPath;
+                language = this.Language;
+                hasAllInformation = this.LastDesignTimeBuildSucceeded;
+            }
+
             var info = ProjectInfo.Create(
```

Comment:
This call is cheap. It might be more sane to read if we just keep all of this under the lock.